cmd/snap-confine/snap-confine.c

```diff
@@ -345,18 +345,17 @@ int main(int argc, char **argv)
 	debug("rgid: %d, egid: %d, sgid: %d",
 	      real_gid, effective_gid, saved_gid);
 
-	// snap-confine runs as both setuid root and setgid root.
-	// Temporarily drop group privileges here and reraise later
-	// as needed.
-	if (effective_gid == 0 && real_gid != 0) {
-		if (setegid(real_gid) != 0) {
-			die("cannot set effective group id to %d", real_gid);
-		}
-	}
-	// This code always needs to run as root for the cgroup/udev setup.
-	if (geteuid() != 0) {
+	// snap-confine needs to run as root for cgroup/udev/mount/apparmor/etc setup.
+	if (effective_uid != 0) {
 		die("need to run as root or suid");
 	}
+	// Make the effective group "root". This ensures that various files we
+	// create are not accidentally group-owned by the user invoking
+	// snap-confine.
+	if (setegid(0) != 0) {
+		die("setegid(0) failed");
+	}
```

Comment:
While I appreciate layered security aspect elsewhere it has an additional cost in complexity. We didn't notice that we are creating files as the wrong group because it wasn't apparent what the state of the process is at each particular time during execution and nobody raised this in code review when some of that code was either introduced or modified (arguably part of the bug is as old as snap-confine itself, the device cgroup is created with wrong group).

I'm not a security expert but the value of constant raise/drop of group doesn't seem like terribly valuable security aspect. It must be handled correctly ahead of every piece of code (including libraries like libapparmor, libseccomp, libcap and others we may link with and invoke) that _may_ create a file or directory.

If you feel strongly about this and overrule this direction then I will adjust each part of snap-confine that handles IO to do the group dance as appropriate. The extra spread test will at least partially make us aware of any new code that is not handling this correctly, with the limitation that does not perform exhaustive execution of all code paths in snap-confine.

Reply:
As mentioned, the code in the PR is too privileged and the solution to dir/file creation isn't to run with more privileges, but to run with minimum privileges needed to perform the action. The way I see it, there are two choices:

1. modify this PR as I suggested and adjust to raise/lower in the couple of places where it is needed
2. modify this PR to `setgid(euid)` when `euid != 0` (ie, permanently drop) and then adjust places where you create files/directories to `chown(path, -1, 0)` unconditionally

Both will achieve the objective of fixing the file/dir creation ownership without running with more privileges than today.

Reply:
>     2\. modify this PR to `setgid(euid)` when `euid != 0` (ie, permanently drop) and then adjust places where you create files/directories to `chown(path, -1, 0)` unconditionally

Correction: there is a small race with this that isn't important with read-only group, but I suspect we'd want a general purpose function to do:
```
chmod(path, 0700);
chown(path, -1, 0);
chmod(path, <what you really want>);
```

Reply:
@jdstrand while I understand the idea is to have the minimum privilege possible I questioned how beneficial this particular (gid handling) is to the overall security of snapd. The need to manage this correctly across the stack has real cost.

I'll look at making changes to this after the sprint, trying to make it less difficult to do correctly than now.

Reply:
This came up again over here: https://github.com/snapcore/snapd/pull/7854#discussion_r354997962

To be clear, this is what I am thinking:
```
void sc_mkdir(const char *parent, const char *subdir, const uid_t uid, const gid_t gid, const mode_t mode) {
    // if non-root, raise our gid to root
    uid_t ruid = getuid();
    if (ruid != 0 && setegid(0) != 0) {
        die(...);
    }

    ... safely and race from create the dir. Jumping off point pseudocode:
    fd = open(parent, ...
    if (fd < 0) ...
    // initially mkdir as 0700, chown then chmod to final to avoid races and capability denials
    if (mkdirat (fd, subdir, 0700) != 0...
    sub_fd = openat(subdir, ...
    if (fchown(sub_fd, uid, gid) != 0...
    if (fchmod (sub_fd, mode) != 0...

    // if non-root, drop our gid to non-root
    if (ruid != 0 && setgid(getgid()) != 0) {
        die(...);
    }
}

int main() {
    // permanently drop right off the bat if we are 6755 (ie, make us behave like 4755)
    gid_t rgid = getgid();
    if (geteuid() == 0 && rgid != 0) {
        if (setgid(rgid) != 0) {
            die(...);
        }
    }

    ...
    sc_mkdir(parent, subdir, 0, 0, 0755);  // via main or any other function
    ...
}

Reply:
I'm iterating on this. It will be ready soon.